File: frontend/js/simulator-contracts.js
/**
 * Conver System — 模拟器域契约单一来源（C8 契约深模块，纯常量 + 纯函数，零副作用）
 *
 * 职责：模拟器域事实常量与 file 安全判据从运行视图 / 列表视图的本地副本收敛
 *   为单一来源 —— 静态目录（SIM_DIR）、清单 URL（由目录派生）、加载超时毫秒数
 *   （TIMEOUT_MS）、清单域超时文案（TIMEOUT_REASON，秒数由毫秒数派生）、file
 *   安全判据纯函数（isValidSimulatorFile）。消费方全部 import 本模块，不再各持
 *   字面量副本：
 *     - frontend/js/simulator-view.js （运行视图：SIM_DIR / TIMEOUT_MS /
 *       isValidSimulatorFile 委托；iframe 超时文案保留自身语义但秒数共享派生）
 *     - frontend/js/simulators.js     （列表视图：MANIFEST_URL / TIMEOUT_MS /
 *       TIMEOUT_REASON）
 *     - frontend/tests/simulator-contracts.test.js（契约锁：常量断言 + 判据矩阵）
 *   维护约定：新增 / 修改模拟器域契约常量只改本模块；各消费文件 docstring 的
 *   共享契约段指向本模块（不复制常量定义）。
 *
 * 文案语义分域：TIMEOUT_REASON 是清单域超时文案（列表视图 fetch 守卫消费）；
 *   运行视图 iframe 加载超时文案「加载超时（N 秒未收到响应）」保留自身语义，
 *   不共用 TIMEOUT_REASON —— 两处仅秒数由同一 TIMEOUT_MS 派生（改毫秒数
 *   必联动两处文案秒数）。
 *
 * 硬约束（Node ESM 真实消费者兼容 — 冒烟脚本直 import 先例，save-key-meta
 *   TD-67/68）：模块顶层零 DOM / 零浏览器 API / 零副作用，仅语言内建
 *   （String.prototype）。
 *
 * 协议表面（__all__）：SIM_DIR / MANIFEST_URL / TIMEOUT_MS / TIMEOUT_REASON /
 *   IMPORT_URL / WARNING_LABELS / isValidSimulatorFile。
 */

// ══════════════════════════════════════════════════
// 常量（模拟器域事实 — 单一来源）
// ══════════════════════════════════════════════════

/** 模拟器静态目录（与列表模块 MANIFEST_URL 同源约定；T2 静态托管根挂载覆盖） */
export const SIM_DIR = 'simulators';

/** 清单 URL（由 SIM_DIR 派生，与原列表视图字面量逐字相同 — 改目录只改一处） */
export const MANIFEST_URL = `${SIM_DIR}/manifest.json`;

/** 加载超时守卫时长（spec 建议 15s；运行视图 iframe 守卫与列表清单守卫共用） */
export const TIMEOUT_MS = 15000;

/** 清单域超时文案（spec D1 逐字；秒数由 TIMEOUT_MS 派生，非硬编码 — 改毫秒数必联动秒数） */
export const TIMEOUT_REASON = `模拟器清单加载超时（${TIMEOUT_MS / 1000} 秒未收到响应）`;

/**
 * 导入端点 URL（工单 04；与后端 03 端点契约逐字一致 — POST /api/simulators/import，
 * multipart 字段名 `file`）。单一来源：导入流程只消费本常量，后端改路径只改一处。
 */
export const IMPORT_URL = '/api/simulators/import';

/**
 * AI 生成游戏端点 URL（JSON POST — { description, title? }）。
 * 成功 200：{ ok: true, game: {...}, retries }
 * 校验失败 422：{ ok: false, errors: [{field, message}], suggestion, retries }
 */
export const GENERATE_URL = '/api/simulators/generate';

/**
 * 恶意模式粗筛键集中文文案映射（工单 04；键集锚定后端
 * simulator_store.SUSPICIOUS_PATTERNS 常量单源 — eval / document.cookie /
 * cross-origin-fetch，增删键必联动后端）。导入成功且 warnings 非空时前端
 * 弹警告不拦截，按本映射逐项展示中文文案；未知键（后端新增未联动）兜底
 * 展示原始键名，不炸。
 */
export const WARNING_LABELS = {
    eval: '使用 eval() 动态执行任意代码（同源可调用 API / 读取本地数据）',
    'document.cookie': '读取 document.cookie（可访问本地会话数据）',
    'cross-origin-fetch': '跨域 fetch 请求（可能向外部发送本地数据）',
};

// ══════════════════════════════════════════════════
// 纯函数：file 安全判据（iframe src 注入守卫）
// ══════════════════════════════════════════════════

/**
 * file 字段安全判据：非空字符串且不含路径分隔符 / 百分号编码 / #。
 *
 * iframe src 注入守卫（等价迁移前运行视图内联判据）：file 来自 manifest
 * 第三方数据，防御越界 / 外链 —— 含 `/` `\` 路径分隔符（TD-56：拒绝穿越与
 * 子路径）、`%` 百分号编码（TD-56：单点拒绝整个百分号编码面 — Starlette
 * 遍历防护与 manifest 可信资产为既有兜底，本判定为纵深加固）或 `#`（URL
 * fragment 分隔符 — iframe src 遇 # 请求截断 → 404，per-game CSS href
 * 同理截断）一律拒绝。对象判定与「参数非法：缺少有效的游戏文件」错误文案
 * 留在运行视图层。
 *
 * @param {unknown} file - 游戏条目的 file 字段值
 * @returns {boolean} 非空字符串且不含 / \ % # 为 true；否则 false
 */
export function isValidSimulatorFile(file) {
    return typeof file === 'string' && file !== ''
        && !file.includes('/') && !file.includes('\\')
        && !file.includes('%') && !file.includes('#');
}

// ══════════════════════════════════════════════════
// 协议表面收口（深模块：外部只通过这些符号与 simulator-contracts.js 交互）
// ══════════════════════════════════════════════════

export const __all__ = [
    'SIM_DIR',
    'MANIFEST_URL',
    'TIMEOUT_MS',
    'TIMEOUT_REASON',
    'IMPORT_URL',
    'GENERATE_URL',
    'WARNING_LABELS',
    'isValidSimulatorFile',
];

File: lib/services/simulator/simulator_contracts.dart
/// 模拟器域契约常量单源（F-M5-02）——SIM_DIR / MANIFEST_FILE / TIMEOUT_MS /
/// 就绪轮询上限 / 固定端口 / file 安全判据单一归属。
///
/// 桌面权威源（只读，语义逐字锚点）：`desktop/frontend/js/simulator-contracts.js`
/// （SIM_DIR='simulators' / MANIFEST_URL 派生 / TIMEOUT_MS=15000 /
/// isValidSimulatorFile）及其契约锁 `desktop/frontend/tests/
/// simulator-contracts.test.js`（15 类判据矩阵）。消费方（F-M5-04 注入 /
/// F-M5-06 存档）只经本模块取模拟器域常量与判据，不各持字面量副本。
///
/// 单一归属派生（F-M5-01 汇报纪律「常量单一归属，后续票不得复制字面量」）：
/// 目录名与清单文件名的**字面量**已随 F-M5-01 合入——`simulator_data_dir.dart`
/// 的 [SimulatorDataDir.subdirectory] 与 `seed_service.dart` 的 [manifestFileName]。
/// 本模块按 SIM_DIR / MANIFEST_FILE 语义**别名消费**这两个既有常量（非复制
/// 字面量），杜绝「两份字面量碰巧相等」的静默破约；其余常量（超时 / 轮询 /
/// 端口）字面量在本模块单一归属。
///
/// 协议表面（深模块）：`SimulatorContracts`（常量 + 判据）。
library;

import 'seed_service.dart' show manifestFileName;
import 'simulator_data_dir.dart' show SimulatorDataDir;

/// 模拟器域契约常量与 file 安全判据——单一来源，零副作用。
abstract final class SimulatorContracts {
  /// 模拟器静态目录名（桌面 SIM_DIR 'simulators'）。URL 路由前缀
  /// `/simulators/`、数据目录名共用；字面量归 F-M5-01
  /// [SimulatorDataDir.subdirectory]，本常量按 SIM_DIR 语义派生，不复制。
  static const String simDir = SimulatorDataDir.subdirectory;

  /// 数据目录下清单文件名（桌面 MANIFEST_FILE 'manifest.json'）。种子标记、
  /// 清单路由、存档收集共用；字面量归 F-M5-01 [manifestFileName]，本常量
  /// 按 MANIFEST_FILE 语义派生，不复制。
  static const String manifestFile = manifestFileName;

  /// 加载超时守卫时长（毫秒；桌面 TIMEOUT_MS=15000 逐字）。列表清单加载
  /// 守卫与运行页 load 超时守卫共用（spec §4.2-4，F-M5-03 消费）。
  static const int timeoutMs = 15000;

  /// 注入脚本就绪轮询上限（毫秒；桌面注入脚本就绪轮询 ≤5s 逐字）。预留常量，
  /// 供 F-M5-04 注入票消费（自包含 JS 常量内的目标控件就绪轮询时长上限）。
  static const int scriptReadyPollMs = 5000;

  /// 本地托管固定端口（共识 D5 定案 8642）。回环监听无公网暴露；被占时
  /// 明确错误态 + 重试，**绝不静默换端口**——换端口 = 换 origin = localStorage
  /// 存档全部静默丢失（spec §3-1）。
  static const int defaultPort = 8642;

  /// 同源反代端点前缀（方案 A CORS 修复，T2/T3 单源）。注入脚本把真实
  /// OpenAI 兼容 base URL 改写为本地 server 同源反代地址 `/proxy/<path>`，
  /// 浏览器 fetch 落在本地 server 上由服务端转发（规避 CORS 拦截）。桌面
  /// 对应 `PROXY_PREFIX = '/api/simulators/proxy'`；移动端本地 server 路由
  /// 挂 `/proxy`（T3 读本常量，与模板占位符替换值共用）。
  static const String proxyPrefix = '/proxy';

  /// file 字段安全判据（锚桌面 simulator-contracts.js isValidSimulatorFile 逐字）：
  /// 非空字符串且不含 `/` `\`（路径分隔符，拒绝穿越与子路径）、`%`（百分号
  /// 编码面）、`#`（URL fragment 分隔符）→ true；其余一律 false。
  ///
  /// 入参取 [Object?] 以吸收 manifest 动态数据（json.decode 产出 Map 值可为
  /// 任意类型）：非字符串类型与空串恒 false，与桌面 typeof 检查语义逐字等价。
  static bool isValidSimulatorFile(Object? file) {
    if (file is! String || file.isEmpty) {
      return false;
    }
    return !file.contains('/') &&
        !file.contains(r'\') &&
        !file.contains('%') &&
        !file.contains('#');
  }
}